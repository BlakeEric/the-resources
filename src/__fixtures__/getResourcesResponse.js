const getResourcesResponse = [
  {
    node: {
      id: 'd04da249-50e7-5b18-a992-d530e14831a7',
      frontmatter: {
        title: 'A List Apart',
        format: 'Article',
        focus: 'Front-end',
        skillLevel: ['Beginner','Moderate','Advanced']
      },
      fields: {
        slug: '/a-list-apart/'
      }
    }
  },
  {
    node: {
      id: '77a6be63-d3c7-50cb-bd29-438427560388',
      frontmatter: {
        title: 'Scrimba',
        format: 'Interactive',
        focus: 'Front-end',
        skillLevel: ['Beginner','Moderate']
      },
      fields: {
        slug: '/scrimba/'
      }
    }
  },
  {
    node: {
      id: '28feb8c5-acb2-5638-bdd6-b0457eff8db3',
      frontmatter: {
        title: 'Smashing Magazine',
        format: 'Article',
        focus: 'Multiple',
        skillLevel: ['Beginner','Moderate','Advanced']
      },
      fields: {
        slug: '/smashing-magazine/'
      }
    }
  },
  {
    node: {
      id: '744d8211-8c35-5c83-b8c1-ecb4ab48c0db',
      frontmatter: {
        title: 'The Odin Project',
        format: 'Tutorial',
        focus: 'Multiple',
        skillLevel: ['Beginner','Moderate']
      },
      fields: {
        slug: '/the-odin-project/'
      }
    }
  },
  {
    node: {
      id: '3b19247d-2ddd-558d-b0ce-14ec1972d52f',
      frontmatter: {
        title: 'Watch and Code',
        format: 'Video',
        focus: 'Front-end',
        skillLevel: ['Beginner']
      },
      fields: {
        slug: '/watch-and-code/'
      }
    }
  }
]

export default getResourcesResponse;
