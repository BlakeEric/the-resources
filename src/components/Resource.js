import React from 'react'
// import { Link } from "gatsby"
import { css } from '@emotion/core'
import { rhythm } from "../utils/typography"

export default function Resource(props) {

  const handleClick = (e) => {
    e.preventDefault();

    if (props.isToggled) {
      props.setToggledItemId(null)
    } else {
      props.setToggledItemId(props.id)
    }
  }


  return (
    <tr key={props.id}>
      <td>
        <a href="#resources" onClick={(e) => handleClick(e)}>
          {props.frontmatter.name}
        </a>
        {props.isToggled &&
          <div css={css`position: relative`}>
            <aside className="resourceDetails" css={resourceDetailStyles}>
              <h4>{props.frontmatter.name}</h4>
              <div dangerouslySetInnerHTML={{ __html: props.html }} />
              <a href={props.frontmatter.url}>View Now</a>
            </aside>
          </div>
        }
      </td>
      <td>
        {props.frontmatter.format}
      </td>
      <td>
        {props.frontmatter.focus}
      </td>
      <td>
        {props.frontmatter.skillLevel.length === 3 ? 'All' : props.frontmatter.skillLevel.join(', ')}
      </td>
    </tr>

  )
}

Resource.defaultProps = {
  isToggled: false,
  setToggledItemId: () => {}
}

const resourceDetailStyles = css`
  position: absolute;
  left: 0;
  max-width: 90vw;
  width: 300px;
  top: 5px;
  left: -${rhythm(.25)};
  background: white;
  border-radius: 5px;
  box-shadow: 2px 2px 12px 1px rgba(0,0,0,0.25);
  padding: ${rhythm(.25)};
  z-index: 10000;
  border: 1px solid #e4e4e4;
  &:after {
    content: "";
    height: 12px;
    width: 12px;
    position: absolute;
    top: 0;
    left: 0;
    background: white;
    transform: rotate(-45deg);
    transform-origin: 0 0;
    left: ${rhythm(1)};
    border-right: 1px solid #e4e4e4;
    border-top: 1px solid #e4e4e4;
  }
  h4 {
    margin-bottom: ${rhythm(.25)};
  }
  p {
    color: black;
  }
  p:last-child {
    margin: 0;
  }
`
