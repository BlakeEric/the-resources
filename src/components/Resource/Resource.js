import React from 'react'
import { css } from '@emotion/core'
import { rhythm } from "../../utils/typography"


export default class Resource extends React.Component {


  componentDidUpdate() {
    if(this.props.isToggled) {
      document.addEventListener('mousedown', this.handleClickOutside);
    } else {
      document.removeEventListener('mousedown', this.handleClickOutside);
    }
  }

  componentWillUnmount() {
    document.removeEventListener('mousedown', this.handleClickOutside);
  }

  /**
   * Set the wrapper ref
   */
  setWrapperRef = (node) => {
    this.wrapperRef = node;
  }

  /**
   * Set Trigger ref
   */
  setTriggerRef = (node) => {
    this.triggerRef = node;
  }

  /**
   * Close detail tab if clicked on outside
   */
  handleClickOutside = (event) => {
    if (
      this.wrapperRef &&
      !this.wrapperRef.contains(event.target) &&
      !this.triggerRef.contains(event.target)
    ) {
      this.props.setToggledItemId(null)
    }
  }

  /**
   * Open detail tab is currently closed, and vice versa
   */
  handleToggle = (e) => {

    e.preventDefault();

    if (this.props.isToggled) {
      this.props.setToggledItemId(null)
    } else {
      this.props.setToggledItemId(this.props.id)
    }
  }

  render() {
    return (
      <tr key={this.props.id} style={{overflow: "visible"}}>
        <td>
        <div className="resourceDetails-wrapper" css={resourceDetailStyles}>
          <a
            href={this.props.frontmatter.url}
            onClick={(e) => this.handleToggle(e)}
            target="_blank" rel="noreferrer noopener"
            ref={this.setTriggerRef}
          >
            {this.props.frontmatter.name}
          </a>
          {this.props.isToggled &&

              <aside
                className={`resourceDetails ${this.props.showDetailsAbove ? 'up' : ''}`}
                ref={this.setWrapperRef}
              >
                <h4>{this.props.frontmatter.name}</h4>
                <div dangerouslySetInnerHTML={{ __html: this.props.html }} />
                <a href={this.props.frontmatter.url} target="_blank" rel="noreferrer noopener">View Now</a>
              </aside>

          }
          </div>
        </td>
        <td>
          {this.props.frontmatter.format}
        </td>
        <td>
          {this.props.frontmatter.focus}
        </td>
        <td>
          {this.props.frontmatter.skillLevel.length === 3 ? 'All' : this.props.frontmatter.skillLevel.join(', ')}
        </td>
      </tr>

    )
  }

}

Resource.defaultProps = {
  isToggled: false,
  setToggledItemId: () => {}
}

const resourceDetailStyles = css`
  position: relative;
  .resourceDetails {
    position: absolute;
    left: 0;
    max-width: 90vw;
    width: 300px;
    top: 110%;
    left: -${rhythm(.25)};
    background: white;
    border-radius: 5px;
    box-shadow: 2px 2px 12px 1px rgba(0,0,0,0.25);
    padding: ${rhythm(.25)};
    z-index: 1000;
    border: 1px solid #e4e4e4;
    &:after {
      content: "";
      height: 12px;
      width: 12px;
      position: absolute;
      top: 0;
      background: white;
      transform: rotate(-45deg);
      transform-origin: 0 0;
      left: ${rhythm(.5)};
      border-right: 1px solid #e4e4e4;
      border-top: 1px solid #e4e4e4;
    }
    &.up {
      top: auto;
      bottom: 110%;
      &:after {
        top: auto;
        bottom: 0;
        transform: rotate(45deg);
        transform-origin: 0 100%;
        border-right: 1px solid #e4e4e4;
        border-bottom: 1px solid #e4e4e4;
        border-top: 0;
      }
    }
    h4 {
      margin-bottom: ${rhythm(.25)};
    }
    p {
      color: black;
    }
    p:last-child {
      margin: 0;
    }
  }
`
