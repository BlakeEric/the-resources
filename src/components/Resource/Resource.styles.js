import { css } from '@emotion/core'
import { rhythm } from "../../utils/typography"

export const resourceDetailStyles = css`
  position: relative;
  .resourceDetails {
    position: absolute;
    left: 0;
    max-width: 90vw;
    width: 300px;
    top: 110%;
    left: -${rhythm(.25)};
    background: white;
    border-radius: 5px;
    box-shadow: 2px 2px 12px 1px rgba(0,0,0,0.25);
    padding: ${rhythm(.5)};
    z-index: 1000;
    border: 1px solid #e4e4e4;
    &:after {
      content: "";
      height: 12px;
      width: 12px;
      position: absolute;
      top: 0;
      background: white;
      transform: rotate(-45deg);
      transform-origin: 0 0;
      left: ${rhythm(.5)};
      border-right: 1px solid #e4e4e4;
      border-top: 1px solid #e4e4e4;
    }
    &.up {
      top: auto;
      bottom: 110%;
      &:after {
        top: auto;
        bottom: 0;
        transform: rotate(45deg);
        transform-origin: 0 100%;
        border-right: 1px solid #e4e4e4;
        border-bottom: 1px solid #e4e4e4;
        border-top: 0;
      }
    }
    h4 {
      margin: 0;
    }
    p {
      color: black;
    }
    p:last-child {
      margin-bottom: ${rhythm(.5)};
    }
    .btn-view {
      padding-right: ${rhythm(.25)};
    }
    button, span {
      display: inline-block;
      font-size: 13px;
    }
    .right {
        float: right;
    }
  }
`