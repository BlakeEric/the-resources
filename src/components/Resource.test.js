import React from "react"
import renderer from "react-test-renderer"
import { shallow, mount } from 'enzyme'
import toJson from 'enzyme-to-json'
import getResourcesResponse from '../__fixtures__/getResourcesResponse'

import Resource from "./Resource"


const mockProps = {
  ...getResourcesResponse[0].node,
  isToggled: false,
  setToggledItemId: () => {}
}

describe("Resource", () => {

  /*
  * Make sure the component renders without error
  */
  it("renders correctly not toggled", () => {
    const tree = renderer
      .create(<Resource {...mockProps}/>)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it("renders correctly when toggled", () => {

    mockProps.isToggled = true;

    const tree = renderer
      .create(<Resource {...mockProps} />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

});
