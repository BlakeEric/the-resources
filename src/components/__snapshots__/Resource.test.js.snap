// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Resource renders correctly not toggled 1`] = `
<tr
  style={
    Object {
      "overflow": "visible",
    }
  }
>
  <td>
    <a
      onClick={[Function]}
      rel="noreferrer noopener"
      target="_blank"
    >
      Alligator
    </a>
  </td>
  <td>
    Article
  </td>
  <td>
    Front-end
  </td>
  <td>
    All
  </td>
</tr>
`;

exports[`Resource renders correctly when toggled 1`] = `
<tr
  style={
    Object {
      "overflow": "visible",
    }
  }
>
  <td>
    <a
      onClick={[Function]}
      rel="noreferrer noopener"
      target="_blank"
    >
      Alligator
    </a>
    <div
      className="resourceDetails-wrapper"
      css={
        Object {
          "map": undefined,
          "name": "402duj",
          "next": undefined,
          "styles": "
  position: relative;
  .resourceDetails {
    position: absolute;
    left: 0;
    max-width: 90vw;
    width: 300px;
    top: 5px;
    left: -0.4025rem;
    background: white;
    border-radius: 5px;
    box-shadow: 2px 2px 12px 1px rgba(0,0,0,0.25);
    padding: 0.4025rem;
    z-index: 1000;
    border: 1px solid #e4e4e4;
    &:after {
      content: \\"\\";
      height: 12px;
      width: 12px;
      position: absolute;
      top: 0;
      left: 0;
      background: white;
      transform: rotate(-45deg);
      transform-origin: 0 0;
      left: 1.61rem;
      border-right: 1px solid #e4e4e4;
      border-top: 1px solid #e4e4e4;
    }
    h4 {
      margin-bottom: 0.4025rem;
    }
    p {
      color: black;
    }
    p:last-child {
      margin: 0;
    }
  }
",
          "toString": [Function],
        }
      }
    >
      <aside
        className="resourceDetails"
      >
        <h4>
          Alligator
        </h4>
        <div
          dangerouslySetInnerHTML={
            Object {
              "__html": undefined,
            }
          }
        />
        <a
          rel="noreferrer noopener"
          target="_blank"
        >
          View Now
        </a>
      </aside>
    </div>
  </td>
  <td>
    Article
  </td>
  <td>
    Front-end
  </td>
  <td>
    All
  </td>
</tr>
`;
